File: code/read/jdbc/src/main/java/be/rubus/microstream/performance/jdbc/query/PurchaseTimeRangeQuery.java
package be.rubus.microstream.performance.jdbc.query;

import be.rubus.microstream.performance.Range;
import be.rubus.microstream.performance.jdbc.query.framework.AbstractJDBCQuery;

import java.sql.*;

public class PurchaseTimeRangeQuery extends AbstractJDBCQuery {

    public Range<Integer> performQuery(Connection connection) {
        Range<Integer> result = null;

        String sql = "SELECT min(p.time_stamp), max(p.time_stamp) FROM purchase p";

        try (PreparedStatement pstmt = connection.prepareStatement(sql)) {
            try (ResultSet resultSet = pstmt.executeQuery()) {
                while (resultSet.next()) {
                    Timestamp min = resultSet.getTimestamp(1);
                    Timestamp max = resultSet.getTimestamp(2);

                    result = new Range<>(min.toLocalDateTime().getYear(), max.toLocalDateTime().getYear());
                }
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return result;
    }

    @Override
    public <T> T extractInstance(Class<T> aClass, ResultSet resultSet, Object extractMetadata) {
        throw new IllegalArgumentException("No data extract required for PurchaseTimeRangeQuery");
    }


}

File: code/read/exposed/src/main/kotlin/be/rubus/microstream/performance/exposed/query/PurchaseTimeRangeQuery.kt
package be.rubus.microstream.performance.exposed.query

import be.rubus.microstream.performance.Range
import be.rubus.microstream.performance.exposed.domain.PurchaseTable
import org.jetbrains.exposed.sql.max
import org.jetbrains.exposed.sql.min
import org.jetbrains.exposed.sql.selectAll
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId

class PurchaseTimeRangeQuery {
    fun performQuery(): Range<Int> {
        val (minTimestamp, maxTimestamp) = PurchaseTable
            .slice(PurchaseTable.timeStamp.min(), PurchaseTable.timeStamp.max())
            .selectAll()
            .map { row ->
                val minTimestamp = row[PurchaseTable.timeStamp.min()]
                val maxTimestamp = row[PurchaseTable.timeStamp.max()]
                Pair(minTimestamp, maxTimestamp)
            }
            .first()

        return Range(getYear(minTimestamp), getYear(maxTimestamp))
    }

    private fun getYear(minTimestamp: Instant?) = LocalDateTime.ofInstant(minTimestamp, ZoneId.systemDefault()).year
}